src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

```diff
@@ -25,7 +25,15 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             }
 
             // Required tokens (order matters)
-            var descriptorKind = reader.ReadNextStringProperty(nameof(TagHelperDescriptor.Kind));
+            var (hash, descriptorKind) = reader.ReadPropertyAndHash(nameof(TagHelperDescriptor.Kind));
```

Comment:
Why can't we just  read the hash and dot he  descriptorKind after  we check the cache? Feels  odd  otherwise

Reply:
If we try to read the hash and fail (which is possible in case they have the "old" format without hash), then the reader would be in a bad state if we used the existing extension methods. Hence, added the `ReadPropertyAndHash` to ensure we don't corrupt reader state in such cases.

One thing this brings up is backwards compatibility. After using this version of the converter, the `project.razor.json` will contain the hashes as well. This is of note because if people use the existing version of the reader then `reader.ReadNextStringProperty` will throw an error because the `HashCode` property isn't recognized.

Reply:
Hmm, can we not look at the property name and if it's unexpected bail out and assume there's no hash? Just asking because the current API feels really odd

Reply:
By looking at the property name we'd perform 1 read. 
https://github.com/dotnet/aspnetcore-tooling/pull/2307/files/2258a0dcc91faa587eeb496ebebe83a660dc8eb3#diff-1e2f5e2505f79b1db576f6c0064e085dL39

If we subsequently bail out, we would re-perform that read in subsequent attempts to read that property name and consequently we'd passover the property being read. 

I've updated the JsonReader extensions to be more fault tolerant for this, however it's worth mentioning this is a bit fragile and could lead to potential issues as we use these serialization functions quite extensively throughout. No issues are being caught by our tests in that regard.